tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/TypesTest.java

```diff
@@ -0,0 +1,345 @@
+        model = project.currentPackage().getCompilation().getSemanticModel(module.moduleId());
+        types = model.types();
+        String org = module.descriptor().org().value();
+        String pkgName = module.descriptor().name().toString();
+        String version = module.descriptor().version().toString();
+
+        Optional<Symbol> symbol = types.getTypeByName(org, pkgName, version, typeDefName);
+        assertTrue(symbol.isPresent());
+        Optional<TypeSymbol> typeSymbol = getTypeSymbolOfTypeDef(symbol.get());
+        assertTrue(typeSymbol.isPresent());
+
+        assertTrue(typeSymbol.get().subtypeOf(expType));
+    }
+
+    @DataProvider(name = "subTypeOfProvider")
+    private Object[][] getSubTypeOfTypes() {
+        return new Object[][] {
+                // foo module
+                {"foo", "MyErr", types.ERROR},
+                {"foo", "ErrorDetail1", types.ANY},
+                {"foo", "MyInt", types.ANY},
+                {"foo", "Language", types.ANY},
+                {"foo", "SI", types.ANY},
+                // bar module
+                {"bar", "MyFloat", types.ANY},
+                {"bar", "Employee", types.ANY},
+                {"bar", "Digit", types.ANY},
+                {"bar", "ABC", types.ANY},
+                {"bar", "PersonClass", types.ANY},
+        };
+    }
+
+    private void testTypesInModule(String pkgName, Object[][] expTypes) {
+        Optional<Map<String, Symbol>> typesInModule = types.typesInModule("testorg", pkgName, "1.0.0");
+        assertTrue(typesInModule.isPresent());
+
+        Map<String, Symbol> typeDefSymbolMap = typesInModule.get();
+        assertEquals(typeDefSymbolMap.size(), expTypes.length);
+
+        for (Object[] expType : expTypes) {
+            String expTypeName = String.valueOf(expType[0]);
+            Symbol symbol = typeDefSymbolMap.get(expTypeName);
+            assertNotNull(symbol);
+            assertSymbolTypeDesc(symbol, (SymbolKind) expType[1], (TypeDescKind) expType[2]);
+        }
+    }
+
+    private void assertSymbolTypeDesc(Symbol symbol, SymbolKind symbolKind, TypeDescKind typeDescKind) {
+        assertEquals(symbol.kind(), symbolKind);
+        Optional<TypeSymbol> typeSymbol = getTypeSymbolOfTypeDef(symbol);
+        assertTrue(typeSymbol.isPresent());
+        assertEquals(typeSymbol.get().typeKind(), typeDescKind);
+    }
+
+    private Optional<TypeSymbol> getTypeSymbolOfTypeDef(Symbol symbol) {
+        switch (symbol.kind()) {
+            case TYPE_DEFINITION:
+                return Optional.of(((BallerinaTypeDefinitionSymbol) symbol).typeDescriptor());
+            case CONSTANT:
+            case ENUM_MEMBER:
+                return Optional.of(((BallerinaConstantSymbol) symbol).typeDescriptor());
+            case ENUM:
+                return Optional.of(((BallerinaEnumSymbol) symbol).typeDescriptor());
+            case CLASS:
+                return Optional.of((BallerinaClassSymbol) symbol);
```

Comment:
Should not be relying on the implementation classes of the symbols unless there's a very good reason for it. Always use the interfaces, even in tests.